CUDD: Reference counting, automatic reordering, node counting

rtjoa opened the pull request:
# Automatic reordering
A reordering type can now be passed into `Cudd`:
```julia
pr(x, algo=Cudd(CUDD.CUDD_REORDER_SIFT))
```

Documentation of reordering types can be found [here](https://add-lib.scce.info/assets/documents/cudd-manual.pdf).

# Node counting
To track the number of nodes, we can pass in a reference to a `CuddDebugInfo`:
```julia
debug_info_ref = Ref{CuddDebugInfo}()
pr(x, algo=Cudd(CUDD.CUDD_REORDER_SIFT, debug_info_ref))
debug_info_ref[].num_nodes  # 122
```

This is admittedly rather ugly. For now, I prioritized making minimal changes and efficiency (only collect debug info when asked), but some other interfaces to consider are:
* Returning a struct instead of a dict (could get messy with multiple queries)
* Making compiling a computation graph to a BDD and BDD inference two separate steps again (`compile` and `infer`), so that we could optionally observe the intermediary (the new version seems cleaner + more efficient though)

# Caveats
`CUDD_REORDER_NONE` and `CUDD_REORDER_EXACT` cause an `EXCEPTION_ACCESS_VIOLATION` for some reason. `CUDD_REORDER_NONE` is not a problem because we can special case it and simply not enable reordering.

For completeness, these are the reordering types that seem to work:
```
CUDD_REORDER_SAME
CUDD_REORDER_RANDOM
CUDD_REORDER_RANDOM_PIVOT
CUDD_REORDER_SIFT
CUDD_REORDER_SIFT_CONVERGE
CUDD_REORDER_SYMM_SIFT
CUDD_REORDER_SYMM_SIFT_CONV
CUDD_REORDER_GROUP_SIFT
CUDD_REORDER_GROUP_SIFT_CONV
CUDD_REORDER_WINDOW2
CUDD_REORDER_WINDOW3
CUDD_REORDER_WINDOW4
CUDD_REORDER_WINDOW2_CONV
CUDD_REORDER_WINDOW3_CONV
CUDD_REORDER_WINDOW4_CONV
CUDD_REORDER_ANNEALING
CUDD_REORDER_GENETIC
```

guyvdbroeck commented:
Would it be possible to check the effect on some of the harder benchmarks currently in the repo?

rtjoa commented:
Just added proper (de)referencing with CUDD.

We see it does improve upon automatic reordering alone from the new test (`Cudd auto reordering`) in `inference_test.jl`:
```
reordering type ┃ num nodes wo deref ┃ num nodes w deref
━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━
none            ┃             196604 ┃            196604
sift            ┃               1500 ┃               160
window2         ┃             196860 ┃            129820
```

@guyvdbroeck - As we just discussed, unfortunately I didn't see any improvements on benchmarks (no change, except `kalman.jl` went from 1126 nodes to 1138). I'll try purposefully introducing some bad orderings to see if it can fix them

I also deleted the inner constructors of `DistAnd`/`DistOr` (which ordered children based on hash) since made it CGs nondeterministic. I updated `canonicalize` so it could still take advantage of and/or's commutativity.
